### PR TITLE
Bump python 3.12 and maintenance updates

### PR DESCRIPTION
**Pull request by ghukill**
### What does this PR do?

Bumps python version to 3.12, and updates structure of project linting and testing to [new conventions](https://mitlibraries.atlassian.net/wiki/spaces/IN/pages/3445948428/DataEng+Application+Maintenance).

Most changes were syntactical to meet new linting requirements, but a couple of repeated changes are somewhat meaningful:
  * all `datetime` objects are now UTC timezone aware
    * anywhere `datetime.datetime(...)` was used, now includes `tzinfo=datetime.UTC`
    * anywhere `datetime.strptime(...)` is used, `.replace(tzinfo=datetime.UTC)` is included to **replace** the timezone with UTC, but keep the original 00:00 hours/minutes when parsing the `YYYY-MM-DD` date string without hours or minutes present
      * if `tzinfo=...` was used, this would bump the hours +4 hours to UTC time upon parsing, which while likely inconsequential, is not ideal, given the `YYYY-MM-DD` timestamp has no notion of hours or minutes

NOTE: Testing has not yet been confirmed in Dev1 or Stage.  This will happen with help from @adamshire123 before a release is tagged.

### Helpful background context

This project was using python 3.11 before this PR, and had not receieved any maintenance updates to match new patterns for linting and testing established recently.

### How can a reviewer manually see the effects of these changes?

Tests run, linting passes.

As mentioned above, manual integration tests that require access to Terraform and Alma console will be performed with help from @adamshire123 (or someone else with access).

### Includes new or updated dependencies?

YES

### What are the relevant tickets?

None

### Developer

- [X] All new ENV is documented in README (or there is none)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -15,23 +15,22 @@ decorator = "*"
...
 boto3-stubs = "*"
 coverage = "*"
 coveralls = "*"
 mock-ssh-server = "*"
 moto = "*"
 mypy = "*"
-pylama = {extras = ["all"], version = "*"}
-pytest = "*"
+pre-commit = "*"
 pydocstyle = "*"
```
Comment (ehanson8) [inline Pipfile]: I don't think we need `pydocstyle` anymore since that's covered by `ruff`

File: sapinvoices/config.py [hunk 2]
```diff
@@ -18,9 +18,10 @@
 
 def configure_logger(logger: logging.Logger, log_level_string: str) -> str:
     if log_level_string.upper() not in logging.getLevelNamesMapping():
-        raise ValueError(f"'{log_level_string}' is not a valid Python logging level")
+        message = f"'{log_level_string}' is not a valid Python logging level"
+        raise ValueError(message)
     log_level = logging.getLevelName(log_level_string.upper())
-    if log_level < 20:
+    if log_level < 20:  # noqa: PLR2004
```
Comment (ehanson8) [inline sapinvoices/config.py]: I've been replacing these with `logging.INFO` and `logging.DEBUG` for clarity (and avoiding that `ruff` error)

File: tests/test_alma.py [hunk 3]
```diff
@@ -168,13 +170,13 @@ def test_mark_invoice_paid_request_status_error(alma_client):
...
     invoice_id = "558809630001021"
-    payment_date = datetime.datetime(2021, 7, 22)
+    payment_date = datetime.datetime(2021, 7, 22, tzinfo=datetime.UTC)
     payment_amount = "120"
     payment_currency = "USD"
     mocked_response = {"payment": {"payment_status": {"value": "FOO"}}}
     with requests_mock.Mocker(case_sensitive=True) as mocker:
         mocker.post(test_url, json=mocked_response)
-        with pytest.raises(ValueError):
+        with pytest.raises(ValueError):  # noqa: PT011
```
Comment (ehanson8) [inline tests/test_alma.py]: Would be nice to get something for the `match=` param but not a dealbreaker 🙂

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+default_language_version:
+  python: python3.11 # set for project python version
```
Comment (jonavellecuerdo) [inline .pre-commit-config.yaml]: Set to `python3.12`

File: Makefile [hunk 5]
```diff
@@ -9,47 +9,54 @@ ECR_URL_DEV:=222053980223.dkr.ecr.us-east-1.amazonaws.com/alma-sapinvoices-dev
 SHELL=/bin/bash
 DATETIME:=$(shell date -u +%Y%m%dT%H%M%SZ)
 
-### Dependency commands ###
+help: # preview Makefile commands
+	@awk 'BEGIN { FS = ":.*#"; print "Usage:  make <target>\n\nTargets:" } \
+/^[-_[:alpha:]]+:.?*#/ { printf "  %-15s%s\n", $$1, $$2 }' $(MAKEFILE_LIST)
 
-install: ## Install dependencies and CLI app
+## ---- Dependency commands ---- ##
```
Comment (jonavellecuerdo) [inline Makefile]: Minor, but might be nice to update: I started using the following syntax for ["section header comments" in the Makefile](https://github.com/MITLibraries/python-cli-template/blob/main/Makefile) to: 

```
###########
# <Section>
###########
```

File: tests/test_sap.py [hunk 6]
```diff
@@ -247,19 +247,19 @@ def test_address_lines_from_address_none_present():
 def test_country_code_from_address_code_present():
     address = {"country": {"value": "USA"}}
     code = sap.country_code_from_address(address)
-    assert "US" == code
+    assert code == "US"
```
Comment (ehanson8) [inline tests/test_sap.py:250]: Ha, now I see where the Yoda conditions were discovered!

**Review by ehanson8 — APPROVED** [review]
Looks great! Some minor suggestions

File: sapinvoices/config.py (hunk 2 above)
Comment (jonavellecuerdo) [inline sapinvoices/config.py]: Change `20` -> `logging.INFO`

File: sapinvoices/sap.py [hunk 7]
```diff
@@ -2,15 +2,15 @@
...
-from datetime import datetime
 from io import StringIO
 from math import fsum
-from typing import Any, List, Literal, Optional, Tuple
+from typing import Any, Literal
 
-import fabric
-import flatdict  # type: ignore
+import fabric  # type: ignore[import-untyped]
+import flatdict  # type: ignore[import-untyped]
```
Comment (jonavellecuerdo) [inline sapinvoices/sap.py]: [Can be ignored via the `pyproject.toml`](https://mypy.readthedocs.io/en/stable/config_file.html#example-pyproject-toml): 

```
[[tool.mypy.overrides]]
module = [
    "fabric", 
    "flatdict"
]
ignore_missing_imports = true
```

**Review by jonavellecuerdo — COMMENTED** [review]
Just a few minor comments but looking good!

File: Pipfile (hunk 1 above)
Comment (ghukill) [reply Pipfile]: Good catch, thanks!

File: tests/test_alma.py (hunk 3 above)
Comment (ghukill) [reply tests/test_alma.py]: Agreed as well, thanks for the nudge.  Formerly the code just raised an empty `ValueError`, but this is a good opportunity to update the exception's message.

Now included in the `ValueError` exception:
> `Invoice '{invoice_id}' not marked as 'PAID' in Alma.`

File: sapinvoices/sap.py (hunk 7 above)
Comment (ghukill) [reply sapinvoices/sap.py]: Good call!  Thanks.  For libraries like these, where it feels unlikely we'll have stubs now or in the future, I like it.

**Comment by ghukill** [comment]
FYI @ehanson8 , @jonavellecuerdo - new commits based on code review suggestions.

**Review by ehanson8 — APPROVED** [review]

**Comment by adamshire123** [comment]
Tested with sample data in stage environment. LGTM.

**Review by adamshire123 — APPROVED** [review]

**Review by jonavellecuerdo — APPROVED** [review]

**Pull request merged**